### PR TITLE
Add labelUse loadFontFamily()

### PR DESCRIPTION
**Pull request by Bouh**
https://github.com/4ian/GDevelop/pull/1394#issuecomment-578529253

### CODE REVIEW

File: Extensions/BBText/JsExtension.js [hunk 1]
```diff
@@ -520,10 +520,7 @@ module.exports = {
         .getProperties(this.project)
         .get('fontFamily')
         .getValue();
-      this._pixiObject.textStyles.default.fontFamily = this._pixiResourcesLoader.getFontFamily(
-        this.project,
-        fontFamily
-      );
+      this._pixiObject.textStyles.default.fontFamily = this._pixiResourcesLoader.loadFontFamily(this.project, fontFamily);
```
Comment (4ian) [inline Extensions/BBText/JsExtension.js]: `loadFontFamily` returns a promise containing the result of the loading, which is the name of the font family to be used. You can't affect it to font family, this won't work (you're affecting a Promise, which is an object, to a string).

Instead, look at how it's done in: https://github.com/4ian/GDevelop/blob/master/newIDE/app/src/ObjectsRendering/Renderers/RenderedTextInstance.js

The reason for having a promise is because the IDE has to load the font - it can't return to you the font family immediately.

**Comment by 4ian** [comment]
Looks good! Can you give a try to a few different fonts (choose a font, see if it's displayed, choose another font, see if displayed, clear the font, see if we're back to arial?) before we go ahead with this? Thanks!!

File: Extensions/BBText/JsExtension.js [hunk 2]
```diff
@@ -520,10 +520,17 @@ module.exports = {
         .getProperties(this.project)
         .get('fontFamily')
         .getValue();
-      this._pixiObject.textStyles.default.fontFamily = this._pixiResourcesLoader.getFontFamily(
-        this.project,
-        fontFamily
-      );
+
+      this._pixiResourcesLoader
+        .loadFontFamily(this.project, fontFamily)
```
Comment (4ian) [inline Extensions/BBText/JsExtension.js]: this._project

**Comment by Bouh** [comment]
Font are correctly displayed, but console log is crazy and display at  each frame
image

**Comment by 4ian** [comment]
So two things:
- It's surely because the PixiResourcesLoader should return "Arial" when you request an empty font resource name (empty string: ""). If I'm not mistaken, these error messages are shown only when the font is empty, right?
- We're lucky to have these because actually there is an important optimization to do! The font should be reloaded when it's changed :) To do this, we should do like in the Renderer for the text objects:

```
if (this._fontName !== xxx.getFontName()) {
    //Avoid calling loadFontFamily if the font didn't changed.
    this._fontName = xxx.getFontName();
    this._pixiResourcesLoader
      .loadFontFamily(this._project, xxx.getFontName())
      ...
}
```

see the idea?

**Comment by Bouh** [comment]
Warnings stay if i add 
```
if (this._pixiObject.textStyles.default.fontFamily !== fontFamily){
 this._pixiResourcesLoader.loadFontFamily().then().catch().....
}

```
Because `fontFamily `from the input is empty. And `this._pixiObject.textStyles.default.fontFamily` is set to `Arial`.
I can't remove `Arial` in `bbTextStyles.default.fontFamily` if i remove it all game with default font will be broken. When i remove this `Arial` font change to another font by default, (Times New Roman ?).

I can do this maybe ?
```
 if (this._pixiObject.textStyles.default.fontFamily !== fontFamily && fontFamily != ""){
 this._pixiResourcesLoader.loadFontFamily().then().catch().....
}
```
I'am find my solution is dirty but working :)

**Comment by Bouh** [comment]
Oooh format is not same, so String of font are never equal.

`this._pixiObject.textStyles.default.fontFamily`  give  `indieflower-ttf`
`fontFamily `(from input resource selector) `IndieFlower.ttf`

**Comment by 4ian** [comment]
Double check my example, you must store the font resource name, not the font family :)

**Comment by Bouh** [comment]
It's working like that. I don't know what is your xxx but i can't use getFontName() there is not a text object.
And pixi-multistyle-text define the input as fontFamily, but we need pass it font resource name with gd 🤔.
Can't merge fontFamily and fontName?
If that's it, I don't get it...

File: Extensions/BBText/JsExtension.js [hunk 3]
```diff
@@ -516,14 +516,27 @@ module.exports = {
...
+      if (this._fontName !== fontResourceName) {
+        console.log('load font');
+
+        this._fontName = fontResourceName;
+
+        this._pixiResourcesLoader
+          .loadFontFamily(this._project, fontResourceName)
+          .then(fontFamily => {
+            // Once the font is loaded, we can use the given fontFamily.
+            this._pixiObject.textStyles.default.fontFamily = fontResourceName;
```
Comment (4ian) [inline Extensions/BBText/JsExtension.js]: this._pixiObject.textStyles.default.fontFamily = fontFamily

**Comment by 4ian** [comment]
Right my explanations were lacking of this nuance. So we have two different things and a mistake in the BBText object:

- The font resource name is the name of the resource in GDevelop. That how you refer to resources in GDevelop: using a resource name (usually, it's automatically set to the name of the file of the resource).

- A "font family" is a notion that does not exist in GDevelop itself. It's a CSS/HTML/pixi-multistyle-text notion. A font family is the name of a font that refers to a CSS font (to keep things simple).

And so we have a mistake in BBText object: 
- this field called "font family" is not a font family! It should have been called "fontResourceName".

We should probably in a next PR fix this.

To summarize, the flow is like this:

```
You get the font resource name -> you call GDevelop resource loader, giving it the resource name (-> GDevelop resource loader loads the font for you) -> you get back a font family (a string) -> you pass it to PIXI -> your font is displayed!
```

> fontFamily is then() is useless

No because, you're getting the font family that you assign to the pixi object :)

File: Extensions/BBText/JsExtension.js [hunk 4]
```diff
@@ -516,14 +516,27 @@ module.exports = {
         .getValue();
       this._pixiObject.textStyles.default.fontSize = `${fontSize}px`;
 
-      const fontFamily = this._associatedObject
+      const fontResourceName = this._associatedObject
```
Comment (4ian) [inline Extensions/BBText/JsExtension.js]: This is the correct name 👍

File: Extensions/BBText/JsExtension.js [hunk 5]
```diff
@@ -516,14 +516,27 @@ module.exports = {
...
         .getValue();
-      this._pixiObject.textStyles.default.fontFamily = this._pixiResourcesLoader.getFontFamily(
-        this.project,
-        fontFamily
-      );
+
+      if (this._fontName !== fontResourceName) {
+        console.log('load font');
+
+        this._fontName = fontResourceName;
```
Comment (4ian) [inline Extensions/BBText/JsExtension.js]: We could even rename `this._fontName` to `this._fontResourceName` (to make extra clear what it represents). I'm ok with both

File: Extensions/BBText/JsExtension.js [hunk 6]
```diff
@@ -516,14 +516,27 @@ module.exports = {
...
+
+        this._pixiResourcesLoader
+          .loadFontFamily(this._project, fontResourceName)
+          .then(fontFamily => {
```
Comment (4ian) [inline Extensions/BBText/JsExtension.js]: This is also correct

**Comment by 4ian** [comment]
Looks good now! Let me know if that works for you and we can merge this.

**Comment by Bouh** [comment]
Yes working fine, thank you :)

**Comment by 4ian** [comment]
Thanks for taking care of that :)
